dds/DCPS/XTypes/DynamicData.cpp

```diff
@@ -9,12 +9,100 @@
 
 #include "DynamicTypeMember.h"
 
+#include <dds/CorbaSeq/LongSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/ULongSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/Int8SeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/UInt8SeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/ShortSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/UShortSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/LongLongSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/ULongLongSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/FloatSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/DoubleSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/LongDoubleSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/CharSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/WCharSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/OctetSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/BooleanSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/StringSeqTypeSupportImpl.h>
+#include <dds/CorbaSeq/WStringSeqTypeSupportImpl.h>
+#include <dds/DCPS/SafetyProfileStreams.h>
+#include <dds/DCPS/RestoreOutputStreamState.h>
+
 #include <ace/OS_NS_string.h>
 
 #include <stdexcept>
 
 OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
 
+namespace {
+#if ACE_SIZEOF_LONG_DOUBLE != 16
+inline ostream& operator<<(ostream& os, const ACE_CDR::LongDouble& val)
+{
+  os << ACE_CDR::LongDouble::NativeImpl(val);
+  return os;
+}
+#endif
+
+inline
+std::ostream& hex_value(std::ostream& o, unsigned value, size_t bytes)
+{
+  OpenDDS::DCPS::RestoreOutputStreamState ross(o);
+  o << std::hex << std::setw(bytes * 2) << std::setfill('0') << value;
+  return o;
+}
+
+template <typename CharType>
+unsigned char_value(CharType value)
+{
+  return value;
+}
+
+#if CHAR_MIN < 0
+/*
+ * If char is signed, then it needs to be reinterpreted as unsigned char or
+ * else static casting '\xff' to a 32-bit unsigned int would result in
+ * 0xffffffff because those are both the signed 2's complement forms of -1.
+ */
+template <>
+inline unsigned char_value<char>(char value)
+{
+  return reinterpret_cast<unsigned char&>(value);
+}
+#endif
+
+template <typename CharType>
+std::ostream& char_helper(std::ostream& o, CharType value)
+{
+  switch (value) {
+  case '\'':
+  case '\"':
+  case '\\':
+  case '\?':
+    return o << '\\' << static_cast<char>(value);
+  case '\n':
+    return o << "\\n";
+  case '\t':
+    return o << "\\t";
+  case '\v':
+    return o << "\\v";
+  case '\b':
+    return o << "\\b";
+  case '\r':
+    return o << "\\r";
+  case '\f':
+    return o << "\\f";
+  case '\a':
+    return o << "\\a";
+  }
+  const unsigned cvalue = char_value(value);
+  if (cvalue <= UCHAR_MAX && isprint(cvalue)) {
+    return o << static_cast<char>(value);
+  }
+  return hex_value(o << "\\x", cvalue, sizeof(CharType) == 1 ? 1 : 2);
+}
+}
```

Comment:
Move this into a common header that's shared with `dds_generator.h` to avoid code duplication.